### PR TITLE
add new warp atomic function

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- add warp function (`getLaneId()`, ~~`getWarpId()`~~, ~~`getNumWarps()`~~
  and `warpBroadcast()`
- add atomics (`atomicAllInc()` and `atomicAllExch()`)

To speedup the new atomic functions in warp communication is used.

### CODE REVIEW

File: src/libPMacc/include/nvidia/warp.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#if (__CUDA_ARCH__ >= 130)
+DINLINE uint32_t getWarpId()
+{
+    uint32_t warpId;
+    asm volatile("mov.u32 %0, %warpid;" : "=r" (warpId));
+    return warpId;
+}
+#endif
+
+/** get maximum number of warps
```
Comment (BenjaminW3) [inline src/libPMacc/include/nvidia/warp.hpp]: What is this exactly?
the maximum number of concurrent warps :
* the hardware allows?
* the current kernel is executed with?

If I call ```getWarpId``` in all threads of a block I assume that all threads get a value in the range [0, ```getNumWarps()```]. Is the whole range used or are there unused warp ids?

Comment (psychocoderHPC) [inline src/libPMacc/include/nvidia/warp.hpp]: IMO thats the maximum what the hardware allows.

CUDA PTX Decumentation Cuda 7.0
```
A predefined, read-only special register that returns the maximum number of warp
identifiers.
```

**Comment by psychocoderHPC** [comment]
removed warp function which currently not needed

File: src/libPMacc/include/nvidia/atomic.hpp [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ *   is used
+ * - all participate threads must change the same
+ *   pointer (ptr) else the result is unspecified
+ *
+ * @param ptr pointer to memory (must be the same address for all threads in a block)
+ *
+ * This warp aggregated atomic increment implementation based on
+ * nvidia parallel forall example
+ * http://devblogs.nvidia.com/parallelforall/cuda-pro-tip-optimized-filtering-warp-aggregated-atomics/
+ */
```
Comment (ax3l) [inline src/libPMacc/include/nvidia/atomic.hpp]: Pls add at least the author of the original post and date of the post.
Also you can write him on GitHub and ask for the license, too - or give him a call, he works in Jülich.
https://github.com/canonizer

**Comment by ax3l** [comment]
@psychocoderHPC do we really want to keep `<sm_20` code? we dropped support for that long ago.

**Comment by psychocoderHPC** [comment]
I added the guard for `sm_1x` because of the fact that this is a asm command. Asm is very hardware specific and therefore I think it is a good style to use such commands only with well defined conditions.

**Comment by ax3l** [comment]
that makes sense, I was more referring to the `if (__CUDA_ARCH__ >= 200)` part in `getLaneId`.

nevertheless, for this specific and short, isolated code section it probably does not matter and can be a bit more general (still, this is basically dead code).

**Pull request merged**